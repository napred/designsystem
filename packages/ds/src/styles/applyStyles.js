// @flow

import { css } from 'emotion';
import type { Styler } from './types';
import type { System } from '../context';

export default function applyStyles(
  props: Object,
  styles: Map<string, Styler<any>>,
  system: System,
) {
  const propNames = Object.keys(props);

  return css(
    propNames.reduce((result, propName) => {
      const style = styles.get(propName);

      if (style == null) {
        return result;
      }

      const appliedStyle = style.apply(props, system);

      return [...result, typeof appliedStyle === 'object' ? css(appliedStyle) : appliedStyle];
    }, []),
  );
}
