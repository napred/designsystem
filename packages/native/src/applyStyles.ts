import { IStyler, ITheme } from '@napred/ds';
import transform from 'css-to-react-native';
import { RegisteredStyle, StyleSheet } from 'react-native';
import Stylis from 'stylis';

const ROOT_CLASS_NAME = '$$root';

function createPlugin() {
  const result: Array<[string, string]> = [];

  return function toArrayPlugin(context: number, content: string, selectors: string[]) {
    if (context === -2) {
      // end
      return result;
    }

    if (context === 1) {
      // we are on line
      if (selectors.length === 1 && selectors[0] === ROOT_CLASS_NAME) {
        // we accept only root styles
        // split line by first : so we get the name of property and value
        const separatorIndex = content.indexOf(':');

        result.push([
          content.substring(0, separatorIndex),
          content.substring(separatorIndex + 1).trim(),
        ]);
      }
    }
  };
}

export default function applyStyles(
  /** Props passed to component */
  props: { [key: string]: any },
  /** Design system */
  system: { theme: ITheme; viewport: number },
  /** All collected styles for component */
  styles: Array<IStyler<any>>,
): { default: RegisteredStyle<any> } {
  const parser = new Stylis();
  parser.use(createPlugin());

  // styler can return only style object or css string
  // we need to convert it to array of tuples of property and value
  const tuples = styles.reduce(
    (result, style) => {
      const appliedStyle = style.apply(props, system);

      // convert css string to tuples
      if (typeof appliedStyle === 'string') {
        const rules = parser(ROOT_CLASS_NAME, appliedStyle);

        return [...result, ...rules];
      } else if (
        typeof appliedStyle === 'object' &&
        appliedStyle != null &&
        !Array.isArray(appliedStyle)
      ) {
        // style object to tuples
        return Object.keys(appliedStyle as { [property: string]: string }).reduce(
          (acc, property) => [
            ...acc,
            // convert value to string because css-to-react-native will throw
            // because it will try to call .trim() on value
            [property, appliedStyle[property].toString()] as [string, string],
          ],
          result,
        );
      } else {
        throw new Error(
          `Invalid style result ${typeof appliedStyle}. Only css string or css object is accepted`,
        );
      }
    },
    [] as Array<[string, string]>,
  );

  return StyleSheet.create({
    default: transform(tuples),
  });
}
