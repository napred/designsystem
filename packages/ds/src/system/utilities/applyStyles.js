// @flow

import getDisplayName from 'react-display-name';
// $FlowExpectError
import styled, { isStyledComponent } from 'styled-components';
import type { ComponentType } from 'react';
import createCleanTag from './createCleanTag';
import collectStylePropNames from './collectStylePropNames';
import isSystemComponent from './isSystemComponent';
import { systemStyles } from '../styles';
import { SystemComponent, type StyleFn } from '../types';

export default function applyStyles<T: Object>(
  Component: string | ComponentType<T> | Class<SystemComponent<T>>,
  styles: Array<StyleFn<any>>,
  parentDefaultProps: Object = {},
): [ComponentType<T>, Array<string>] {
  if (isStyledComponent(Component)) {
    throw new Error('U mad fam? No styled component, aight?!');
  }

  let C = Component;
  // Flow hack because systemStyles is array of styler
  // we just use $Compose as a hack to get props of all styles
  let allStyles = [...(systemStyles: any), ...styles];
  let blacklist = collectStylePropNames(styles);
  let defaultProps = {};

  if (typeof Component === 'string') {
    C = createCleanTag(Component);
  } else if (isSystemComponent(Component)) {
    const SysComponent: Class<SystemComponent<any>> = (Component: any);

    if ((SysComponent: any).defaultProps != null) {
      defaultProps = {
        ...(SysComponent: any).defaultProps,
        ...parentDefaultProps,
      };
    }

    C = SysComponent.targetComponent;
    allStyles = [...SysComponent.appliedStyles, ...allStyles];
    blacklist = [...blacklist, ...SysComponent.blacklist];
  } else if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(
      `You are trying to use custom component ${getDisplayName(
        Component,
      )} as base component, it can cause problem with blacklisting, etc`,
    );
  }

  const StyledComponent = styled((C: any))`
    /* prettier-ignore */
    ${allStyles}
  `;

  StyledComponent.defaultProps = defaultProps;

  return [StyledComponent, blacklist];
}
