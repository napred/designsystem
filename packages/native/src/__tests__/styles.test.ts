import { defaultTheme, nativeStyleList } from '../';
import applyStyles from '../applyStyles';

describe('test styles', () => {
  it('works for string styles', () => {
    expect(
      applyStyles(
        {
          alignContent: 'center',
          alignItems: 'center',
          alignSelf: 'center',
          bgColor: '#ccc',
          borderColor: '#ccc',
          borderStyle: 'dashed',
          color: '#ccc',
          display: 'flex',
          flex: '1',
          fontWeight: 'bold',
          overflow: 'hidden',
          overflowX: 'hidden', // not supported by react native, ignored in output
          overflowY: 'hidden', // not supported by react native, ignored in output
          position: 'absolute',
          shadow: '12px 12px 1px rgba(0, 0, 255, .2)',
          textAlign: 'center',
          textOverflow: 'ellipsis', // - not supported by react native, ignored in output
          textTransform: 'underline',
          transition: 'margin-right 2s ease-in-out .5s', // - not supported by react native, ignored in output
          whiteSpace: 'nowrap', // not supported by react native, ignored in output
        },
        { theme: defaultTheme, viewport: 0 },
        nativeStyleList,
      ),
    ).toMatchSnapshot();
  });

  it('works correctly for numeric styles', () => {
    expect(
      applyStyles(
        {
          borderRadius: '10px',
          borderWidth: '10px',
          bottom: '10px',
          flexBasis: '100%',
          flexGrow: 1,
          flexShrink: 1,
          fontWeight: 300,
          height: '10px',
          left: '10px',
          m: 0,
          maxHeight: '10px',
          maxWidth: '10px',
          minHeight: '10px',
          minWidth: '10px',
          opacity: 0.8,
          p: 0,
          right: '10px',
          top: '10px',
          width: '10px',
          zIndex: 10,
        },
        { theme: defaultTheme, viewport: 0 },
        nativeStyleList,
      ),
    ).toMatchSnapshot();
  });
});
