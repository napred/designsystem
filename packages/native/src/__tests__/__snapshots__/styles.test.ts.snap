// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test styles works correctly for numeric styles 1`] = `
Object {
  "default": Object {
    "borderBottomLeftRadius": 10,
    "borderBottomRightRadius": 10,
    "borderBottomWidth": 10,
    "borderLeftWidth": 10,
    "borderRightWidth": 10,
    "borderTopLeftRadius": 10,
    "borderTopRightRadius": 10,
    "borderTopWidth": 10,
    "bottom": 10,
    "flexBasis": "100%",
    "flexGrow": 1,
    "flexShrink": 1,
    "fontFamily": "Palanquin",
    "fontSize": 16,
    "fontWeight": "300",
    "height": 10,
    "left": 10,
    "lineHeight": 24,
    "marginBottom": 0,
    "marginLeft": 0,
    "marginRight": 0,
    "marginTop": 0,
    "maxHeight": 10,
    "maxWidth": 10,
    "minHeight": 10,
    "minWidth": 10,
    "opacity": 0.8,
    "paddingBottom": 0,
    "paddingLeft": 0,
    "paddingRight": 0,
    "paddingTop": 0,
    "right": 10,
    "top": 10,
    "width": 10,
    "zIndex": 10,
  },
}
`;

exports[`test styles works for string styles 1`] = `
Object {
  "default": Object {
    "alignContent": "center",
    "alignItems": "center",
    "alignSelf": "center",
    "backgroundColor": "#ccc",
    "borderBottomColor": "#ccc",
    "borderLeftColor": "#ccc",
    "borderRightColor": "#ccc",
    "borderStyle": "dashed",
    "borderTopColor": "#ccc",
    "color": "#ccc",
    "display": "flex",
    "flexBasis": 0,
    "flexGrow": 1,
    "flexShrink": 1,
    "fontFamily": "Palanquin",
    "fontSize": 16,
    "fontWeight": "bold",
    "lineHeight": 24,
    "overflow": "hidden",
    "position": "absolute",
    "shadowColor": "rgba(0, 0, 255, .2)",
    "shadowOffset": Object {
      "height": 12,
      "width": 12,
    },
    "shadowOpacity": 1,
    "shadowRadius": 1,
    "textAlign": "center",
  },
}
`;
